Arithmetic makes borrow checker overly restrictive

UPDATE: This was fixed by #47167 but still [needs a regression test to be added](https://github.com/rust-lang/rust/issues/29743#issuecomment-358650299).

----

``` rust
fn main() {
    let mut i = [1,2,3];
    i[i[0]] = 0; // Works
    i[i[0] - 1] = 0; // Is rejected
}
```

I had expect the `- 1` not to make any difference, but the borrow checker complains with the following error:

```
<anon>:4:7: 4:11 error: cannot use `i[..]` because it was mutably borrowed
<anon>:4     i[i[0] - 1] = 0;
               ^~~~
<anon>:4:5: 4:6 note: borrow of `i` occurs here
<anon>:4     i[i[0] - 1] = 0;
```

Still seems to be a problem on rustc 1.23.0-nightly (8b22e70b2 2017-10-31) - not sure if expected ?

```
error[E0503]: cannot use `i[..]` because it was mutably borrowed
 --> main.rs:4:7
  |
4 |     i[i[0] - 1] = 0; // Is rejected
  |     - ^^^^ use of borrowed `i`
  |     |
  |     borrow of `i` occurs here

error: aborting due to previous error
```

--- a/src/test/ui/issues/issue-24338.rs
+++ b/src/test/ui/issues/issue-24338.rs
@@ -1,0 +1,30 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+//
+// compile-pass
+
+trait DictLike<'a> {
+    type ItemsIterator: Iterator<Item=u8>;
+    fn get(c: Self::ItemsIterator) {
+        c.into_iter();
+    }
+}
+
+trait DictLike2<'a> {
+    type ItemsIterator: Iterator<Item=u8>;
+
+    fn items(&self) -> Self::ItemsIterator;
+
+    fn get(&self)  {
+        for _ in self.items() {}
+    }
+}
+
+fn main() {}

--- a/src/test/ui/issues/issue-28134.rs
+++ b/src/test/ui/issues/issue-28134.rs
@@ -1,0 +1,13 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// compile-flags: --test
+
+#![test] //~ ERROR only functions may be used as tests

--- a/src/test/ui/issues/issue-28134.stderr
+++ b/src/test/ui/issues/issue-28134.stderr
@@ -1,0 +1,8 @@
+error: only functions may be used as tests
+  --> $DIR/issue-28134.rs:13:1
+   |
+LL | #![test] //~ ERROR only functions may be used as tests
+   | ^^^^^^^^
+
+error: aborting due to previous error
+

--- a/src/test/ui/issues/issue-29743.rs
+++ b/src/test/ui/issues/issue-29743.rs
@@ -1,0 +1,17 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+//
+// compile-pass
+
+fn main() {
+    let mut i = [1, 2, 3];
+    i[i[0]] = 0;
+    i[i[0] - 1] = 0;
+}

